src/message.rs

```diff
@@ -36,14 +36,160 @@ impl Deserialize for Version {
     }
 }
 
+/// An RPC method params
+///
+/// Not all json is a valid RPC params.
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+#[serde(untagged)]
+pub enum Params {
+    Positional(Vec<Value>),
+    Named(serde_json::Map<String, Value>),
+}
+
+impl Params {
+    /// Wether or not there are any params
+    pub fn is_empty(&self) -> bool {
+        match *self {
+            Params::Positional(ref args) => args.is_empty(),
+            Params::Named(ref fields) => fields.is_empty(),
+        }
+    }
+
+    /// Returns a json value consuming the params in the process
+    pub fn into_value(self) -> Value {
+        match self {
+            Params::Positional(xs) => Value::Array(xs),
+            Params::Named(x) => Value::Object(x),
+        }
+    }
+
+    /// Force the params to be represented in positional form
+    pub fn into_positional(self) -> Self {
+        match self {
+            Params::Named(x) => {
+                let values = x.into_iter()
+                    .map(|(_, v)| v)
+                    .collect();
+                Params::Positional(values)
+            },
+            x => x,
+        }
+    }
+}
+
+// Helper fuction for skipping the params field serialization when there are no params
+pub fn params_empty(x: &Option<Params>) -> bool {
+    match *x {
+        Some(ref x) => x.is_empty(),
+        None => true,
+    }
+}
+
+// TODO: how do we sanitize uses of `json_internal!()`? check the rocket crate
+// TODO: missing docs
+#[macro_export]
+macro_rules! params {
+
+    //////////////////////////////////////////////////////////////////////////
+    // The main implementation.
+    //
+    // Must be invoked as: params!($($json)+)
+    //////////////////////////////////////////////////////////////////////////
+
+    () => {
+        None
+    };
+
+    // Review: do we want to use this to send a `"params": null` as opposed to skipping it?
+    //         params!() or None can be the skip while params!(null) will actually send null
+    //         Notes:
+    //           - Params::Null needs to be added
+    //           - `None` and `Some(Params::Null)` are "almost" equivalent on server side
+    (null) => {
+        None
+    };
+
+    (true) => {
+        Some($crate::Params::Positional(vec![$crate::macro_exports::Value::Bool(true)]))
+    };
+
+    (false) => {
+        Some($crate::Params::Positional(vec![$crate::macro_exports::Value::Bool(false)]))
+    };
+
+    ([]) => {
+        {
+            Some($crate::Params::Positional(vec![]))
+        }
+    };
+
+    ([ $($tt:tt)+ ]) => {
+        {
+            match $crate::macro_exports::Value::Array(json_internal!(@array [] $($tt)+)) {
```

Comment:
Wouldn't `match json!([ $($tt)+ ] {` be easier? Or just directly `$crate::Params::Positional(json_internal!(...))`? This construct array and then deconstruct it right away seems a bit strange.